SecurityException: Caller no longer running / JobIntentService

Crash report from users and Google Play:

- Sent: 27.12.2018
- App version: [v.1.38.1](https://github.com/EventFahrplan/EventFahrplan/releases/tag/v.1.38.1), [v.1.38.2](https://github.com/EventFahrplan/EventFahrplan/releases/tag/v.1.38.2), [v.1.40.0.](https://github.com/EventFahrplan/EventFahrplan/releases/tag/v.1.40.0)
- Android version: 8.0.0, 8.1.0, 9
- Device model: various

# Stacktrace Android 8
``` java
java.lang.RuntimeException: An error occurred while executing doInBackground()
  at android.os.AsyncTask$3.done(AsyncTask.java:353)
  at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
  at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
  at java.util.concurrent.FutureTask.run(FutureTask.java:271)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
  at java.lang.Thread.run(Thread.java:764)

Caused by: java.lang.SecurityException: Caller no longer running, last stopped +331ms
  because: timed out while starting
  at android.os.Parcel.readException(Parcel.java:1960)
  at android.os.Parcel.readException(Parcel.java:1899)
  at android.app.job.IJobCallback$Stub$Proxy.dequeueWork(IJobCallback.java:191)
  at android.app.job.JobParameters.dequeueWork(JobParameters.java:196)
  at android.support.v4.app.JobIntentService$JobServiceEngineImpl.dequeueWork(JobIntentService.java:314)
  at android.support.v4.app.JobIntentService.dequeueWork(JobIntentService.java:639)
  at android.support.v4.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:389)
  at android.support.v4.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:382)
  at android.os.AsyncTask$2.call(AsyncTask.java:333)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  ... 3 more
```


# Stacktrace Android 9
``` java
java.lang.RuntimeException: An error occurred while executing doInBackground()
  at android.os.AsyncTask$3.done(AsyncTask.java:354)
  at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
  at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
  at java.util.concurrent.FutureTask.run(FutureTask.java:271)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  at java.lang.Thread.run(Thread.java:764)

Caused by: java.lang.SecurityException: Caller no longer running, last stopped +3s357ms 
  because: timed out while starting
  at android.os.Parcel.createException(Parcel.java:1942)
  at android.os.Parcel.readException(Parcel.java:1910)
  at android.os.Parcel.readException(Parcel.java:1860)
  at android.app.job.IJobCallback$Stub$Proxy.dequeueWork(IJobCallback.java:195)
  at android.app.job.JobParameters.dequeueWork(JobParameters.java:243)
  at android.support.v4.app.JobIntentService$JobServiceEngineImpl.dequeueWork(JobIntentService.java:314)
  at android.support.v4.app.JobIntentService.dequeueWork(JobIntentService.java:639)
  at android.support.v4.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:389)
  at android.support.v4.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:382)
  at android.os.AsyncTask$2.call(AsyncTask.java:333)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  ... 3 more

Caused by: android.os.RemoteException: Remote stack trace:
  at com.android.server.job.JobServiceContext.assertCallerLocked(JobServiceContext.java:481)
  at com.android.server.job.JobServiceContext.doDequeueWork(JobServiceContext.java:359)
  at com.android.server.job.JobServiceContext$JobCallback.dequeueWork(JobServiceContext.java:160)
  at android.app.job.IJobCallback$Stub.onTransact(IJobCallback.java:83)
  at android.os.Binder.execTransact(Binder.java:731)
```

# Log
``` java
 0 D: Writing unhandled exception to: /data/user/0/info.metadude.android.congress.schedule/files/1.38.2-1545918814057.tracedroid
 1 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag="e23f36036636b21b1c92d260eee962deec6fcdc9", numDays=4, subtitle=, title=35th Chaos Communication Congress, version=Golden Girls 3 1.7 2.0.2 1.1 MemoryRefreshV1.0 day1 morning fixing timezone issue 1.0.6 2018-12-27 14:40)
 2 D: viewDay(false)
 3 D: meta.getNumDays() = 4
 4 D: MyApp.task_running = NONE
 5 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag="e23f36036636b21b1c92d260eee962deec6fcdc9", numDays=4, subtitle=, title=35th Chaos Communication Congress, version=Golden Girls 3 1.7 2.0.2 1.1 MemoryRefreshV1.0 day1 morning fixing timezone issue 1.0.6 2018-12-27 14:40)
 6 D: viewDay(false)
 7 D: lecture_id = null
 8 D: MyApp.lectureList contains 251 items.
 9 D: onResume

10 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag="e23f36036636b21b1c92d260eee962deec6fcdc9", numDays=4, subtitle=, title=35th Chaos Communication Congress, version=Golden Girls 3 1.7 2.0.2 1.1 MemoryRefreshV1.0 day1 morning fixing timezone issue 1.0.6 2018-12-27 14:40)
11 D: viewDay(false)
12 D: meta.getNumDays() = 4
13 D: MyApp.task_running = NONE
14 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag="e23f36036636b21b1c92d260eee962deec6fcdc9", numDays=4, subtitle=, title=35th Chaos Communication Congress, version=Golden Girls 3 1.7 2.0.2 1.1 MemoryRefreshV1.0 day1 morning fixing timezone issue 1.0.6 2018-12-27 14:40)
15 D: viewDay(false)
16 D: lecture_id = null
17 D: MyApp.lectureList contains 251 items.
18 D: onResume

19 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag="e23f36036636b21b1c92d260eee962deec6fcdc9", numDays=4, subtitle=, title=35th Chaos Communication Congress, version=Golden Girls 3 1.7 2.0.2 1.1 MemoryRefreshV1.0 day1 morning fixing timezone issue 1.0.6 2018-12-27 14:40)
20 D: viewDay(false)
21 D: meta.getNumDays() = 4
22 D: MyApp.task_running = NONE
23 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag="e23f36036636b21b1c92d260eee962deec6fcdc9", numDays=4, subtitle=, title=35th Chaos Communication Congress, version=Golden Girls 3 1.7 2.0.2 1.1 MemoryRefreshV1.0 day1 morning fixing timezone issue 1.0.6 2018-12-27 14:40)
24 D: viewDay(false)
25 D: lecture_id = null
26 D: MyApp.lectureList contains 251 items.
27 D: onResume

28 D: current handler class=com.android.internal.os.RuntimeInit$KillApplicationHandler
```

# Related
-  [Issue #63622293: JobIntentService is crashing in an AsyncTask on Android O](https://issuetracker.google.com/issues/63622293)

--- a/app/src/main/java/android/support/v4/app/SafeJobIntentService.java
+++ b/app/src/main/java/android/support/v4/app/SafeJobIntentService.java
@@ -1,0 +1,22 @@
+package android.support.v4.app;
+
+import org.ligi.tracedroid.logging.Log;
+
+/**
+ * Mitigates the effect of a SecurityException which can occur in {@link JobIntentService}.
+ * The issue is tracked here: https://issuetracker.google.com/issues/63622293
+ */
+public abstract class SafeJobIntentService extends JobIntentService {
+
+    @Override
+    GenericWorkItem dequeueWork() {
+        try {
+            return super.dequeueWork();
+        } catch (SecurityException ignore) {
+            // There is not much we can do here.
+            Log.e(getClass().getName(), ignore);
+        }
+        return null;
+    }
+
+}

--- a/app/src/main/java/nerd/tuxmobil/fahrplan/congress/autoupdate/UpdateService.java
+++ b/app/src/main/java/nerd/tuxmobil/fahrplan/congress/autoupdate/UpdateService.java
@@ -7,8 +7,8 @@ import android.content.SharedPreferences;
 import android.net.Uri;
 import android.preference.PreferenceManager;
 import android.support.annotation.NonNull;
-import android.support.v4.app.JobIntentService;
 import android.support.v4.app.NotificationCompat;
+import android.support.v4.app.SafeJobIntentService;
 import android.text.TextUtils;
 
 import java.util.List;
@@ -32,7 +32,7 @@ import nerd.tuxmobil.fahrplan.congress.schedule.MainActivity;
 import nerd.tuxmobil.fahrplan.congress.utils.FahrplanMisc;
 import okhttp3.OkHttpClient;
 
-public class UpdateService extends JobIntentService {
+public class UpdateService extends SafeJobIntentService {
 
     private static final int JOB_ID = 2119;
 

